### PR TITLE
Collocate Details::Node implementation details into single header

### PR DESCRIPTION
**Pull request by dalg24**

### CODE REVIEW

File: src/details/ArborX_DetailsNode.hpp [hunk 1]
```diff
@@ -22,19 +22,27 @@ namespace Details
...
-  KOKKOS_INLINE_FUNCTION
-  Node() = default;
+  KOKKOS_FUNCTION
+  constexpr Node() = default;
 
-  KOKKOS_INLINE_FUNCTION
-  Node(const Kokkos::pair<Node *, Node *> &c, const Box &bb)
-      : children(c)
-      , bounding_box(bb)
+  KOKKOS_FUNCTION constexpr bool isLeaf() const noexcept
```
Comment (aprokop) [inline src/details/ArborX_DetailsNode.hpp]: Please change `KOKKOS_FUNCTION` back to `KOKKOS_INLINE_FUNCTION` in all instances.

File: src/details/ArborX_DetailsNode.hpp [hunk 2]
```diff
@@ -22,19 +22,27 @@ namespace Details
...
+  KOKKOS_FUNCTION std::size_t getLeafPermutationIndex() const noexcept
+  {
+    return reinterpret_cast<std::size_t>(children.second);
   }
 
   Kokkos::pair<Node *, Node *> children = {nullptr, nullptr};
   Box bounding_box;
 };
+
+KOKKOS_INLINE_FUNCTION Node makeLeafNode(std::size_t index, Box box) noexcept
```
Comment (aprokop) [inline src/details/ArborX_DetailsNode.hpp]: please rename `index` -> `permutation_index` for clarity

File: src/details/ArborX_DetailsNode.hpp [hunk 3]
```diff
@@ -22,19 +22,27 @@ namespace Details
...
-  Node(const Kokkos::pair<Node *, Node *> &c, const Box &bb)
-      : children(c)
-      , bounding_box(bb)
+  KOKKOS_FUNCTION constexpr bool isLeaf() const noexcept
   {
+    return children.first == nullptr;
+  }
+
+  KOKKOS_FUNCTION std::size_t getLeafPermutationIndex() const noexcept
+  {
```
Comment (dalg24) [inline src/details/ArborX_DetailsNode.hpp]: Could do
```C++
assert(isLeaf());
```
here.  Result is garbage otherwise.

Comment (aprokop) [reply src/details/ArborX_DetailsNode.hpp]: Yeah, I had the same thought, and was not sure. I think it's fine as it is, as it does not change status quo.

File: src/details/ArborX_DetailsNode.hpp (hunk 1 above)
Comment (dalg24) [reply src/details/ArborX_DetailsNode.hpp]: Explicitly defaulted default constructor as well?

Comment (aprokop) [reply src/details/ArborX_DetailsNode.hpp]: Sure, if only for consistency. I think KOKKOS_INLINE_FUNCTION is more robust if at some point we move the body of the constructor/function out of the class. On the other hand, you can argue that you'll remember to switch `KOKKOS_FUNCTION` at that point.

File: src/details/ArborX_DetailsNode.hpp (hunk 3 above)
Comment (dalg24) [reply src/details/ArborX_DetailsNode.hpp]: I decided to add it.

File: src/details/ArborX_DetailsNode.hpp (hunk 1 above)
Comment (dalg24) [reply src/details/ArborX_DetailsNode.hpp]: Done.

File: src/details/ArborX_DetailsNode.hpp (hunk 2 above)
Comment (dalg24) [reply src/details/ArborX_DetailsNode.hpp]: Done.

**Comment by dalg24** [comment]
> Looks good. I highly recommend squashing all commits into one.

Working on my stats :p

**Comment by aprokop** [comment]
> Working on my stats :p

@dalg24 You have 19 commits for 39 lines of code. Indeed. :frowning_face:

**Pull request merged**